src/sentry/roles/manager.py

```diff
@@ -31,55 +51,155 @@ def has_scope(self, scope: str) -> bool:
         return scope in self.scopes
 
 
-class RoleManager:
-    def __init__(
-        self,
-        config: Iterable[Mapping[str, str]],
-        default: Optional[str] = None,
-    ) -> None:
-        role_list = []
-        self._roles: Dict[str, Role] = OrderedDict()
-        for idx, role_cfg in enumerate(config):
-            role = Role(idx, **role_cfg)
-            role_list.append(role)
-            self._roles[role.id] = role
+@dataclass(frozen=True, eq=True)
+class OrganizationRole(Role):
+    is_global: bool = False
+
+    def can_manage(self, other: OrganizationRole) -> bool:
+        return self.priority >= other.priority
+
+    def get_minimum_team_role(self) -> TeamRole:
+        """Return the minimum team role for this organization role.
+
+        A member with this organization role automatically receives the minimum role
+        when joining a new team, and can't be demoted below that team role for as
+        long as they hold the organization role.
+        """
+        return self.parent.get_minimum_team_role(self.id)
+
+    def can_manage_team_role(self, other: TeamRole) -> bool:
+        return self.get_minimum_team_role().can_manage(other)
 
-        self._choices = tuple((r.id, r.name) for r in role_list)
 
-        if default:
-            self._default = self._roles[default]
-        else:
-            self._default = role_list[0]
+@dataclass(frozen=True, eq=True)
+class TeamRole(Role):
+    is_minimum_role_for: str | None = None
 
-        self._top_dog = role_list[-1]
+    def can_manage(self, other: TeamRole) -> bool:
+        return self.priority >= other.priority
 
-    def __iter__(self) -> Iterable[Role]:
-        yield from self._roles.values()
+
+class RoleLevel(Generic[R]):
+    """Represent the set of all roles at one level (org or team)."""
+
+    def __init__(self, roles: Iterable[R], default_id: str | None = None) -> None:
+        self._priority_seq = tuple(sorted(roles, key=lambda r: r.priority))
+        self._id_map = OrderedDict((r.id, r) for r in self._priority_seq)
+
+        self._choices = tuple((r.id, r.name) for r in self._priority_seq)
+        self._default = self._id_map[default_id] if default_id else self._priority_seq[0]
+        self._top_dog = self._priority_seq[-1]
```

Comment:
What does top_dog mean in context of a team level role?

Reply:
Same as for org roles, the one with the highest priority among all team roles.

The definition is useful on [line 165](https://github.com/getsentry/sentry/pull/33387/files#diff-374f5f57d7ec405d494179b0862030ee18102153405145c9ab692dc06d0186ceR165), which ensures that all team roles can be assigned, at minimum, by an org-level top dog.